### DISCUSSION TITLE
Middleware Logic Implementation

### DISCUSSION THREAD

**Discussion by vimalmistry** [uNetworking/uWebSockets.js#483] [Ideas]
I am trying to implement middleware pipeline feature for my project.
previously I tried something ..

```ts
app.get('/x', preload(authorized(IndexController.index)));
```

But I want some clean code and managing those function which returns handle functions are messy for me :(

I am trying to do something like this

```ts
app.get('/', handle([m1,m2,m3], controller));
```

Here is my handle function. Which kinda works. but If I bombard request with benchmark tool or keep refreshing tab on browser. I am getting following error. That's why I used try catch on my all code 😃
I notice that Its occurs when I `res.end` in some off middleware. I am pretty new in node js. Just working in my spare time. I am posting here so you guys can guide me or give me valuable suggestions. Thanks.

Error is

> Listening to port 9001
UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "Invalid access of discarded (invalid, deleted) uWS.HttpResponse/SSLHttpResponse.".

```ts

export default function handle(middlewares: Middleware[], handler: RequestHandler): RequestHandler {
    return async (res: HttpResponse, req: HttpRequest) => {

        res.onAborted(() => res.aborted = true);

        res.json = (data: object) => res.end(JSON.stringify(data));

        try {
            await Promise.all(
                middlewares.map(async (fn) => {
                    try {
                        return await fn(res, req);
                    } catch (error) {
                        console.log('hereeee', error);
                    }
                })
            ).catch((_) => console.log('hereee')
            );

            return handler(res, req);
        } catch (error) {
            return res.writeStatus("500").json({
                error: "Server Errror!"
            })
        }
    };
}

```

Here is Example Controller and Middleware
```ts
const controller: RequestHandler = (res, req) => {
    res.end(res.xx);
}

const m1: Middleware = async (res, req) => {
    res.xx = '500';
    res.writeStatus("500").end('plolad')
}

```

**Comment by hst-m** [comment]
@vimalmistry FYI Promise.all() runs all the promises at the same time, so your code is running all your middleware for the route at same time, probably not what you want

**Reply by vimalmistry** [reply]
Yep. I noticed

**Comment by snkj5** [comment]
U can setup base middleware system smth like this:

```javascript
const middlewares = [
   (res, req, next) => { ... },
   (res, req, next) => { ... }
];

const runMiddlewares = (res, req) => {
    let index = 0;

    const next = async () => {
        const nextMiddleware = middlewares[index++];
        const hasRespondEnded = res.finished; // this is CUSTOM property, see my comment below
        const shouldRunNextMiddleware = !hasRespondEnded && nextMiddleware;

        if (shouldRunNextMiddleware) {
            nextMiddleware(res, req, next);
        }
    };

    next();
};

App = new App().any('/*', runMiddlewares)

```
Unfortunately, today's HttpResponse API doesnt provide easy way to check if response has finished.
For now you propably might want to override the res.end and res.tryEnd methods:

```javascript
const enhanceRes = (res) => {
     const originalEnd = res.end.bind(res);
     const originalTryEnd = res.tryEnd.bind(res);
     
     res.finished = false;
     
     res.end = (...args) => {
         originalEnd(...args);
         res.finished = true;
         return res;
     }
     
     res.tryEnd = (...args) => {
          const [ok, done] = originalTryEnd(...args);
          res.finished = done;
          return [ok, done]
     }
}
```

**Reply by vimalmistry** [reply]
Great. I created similar logic for my middleware. But your code is insanely simple :) .
And Thanks for response override suggestion :)

**Reply by haoynmail** [reply]
hi kolodziejczak-sz:

Please provide guidance
The middleware system you implemented:
     const middlewares = [
           (res, req, next) => { ... },
           (res, req, next) => { ... }
           (res, req, next) => { ... }
        ];

     const runMiddlewares = (res, req) => {...} 
     const enhanceRes = (res) => {...}
     
How to use **_enhanceRes_**
For example: 
     const middlewares = [
           (res, req, next) => { ... },
           (res, req, next) => { ... },
           (res, req, next) => { ... }
        ];

     const app=uWS App ()

     app. get ('/a', (res, req)=>{
         runMiddlewares(res,req); first middleware to run
     }). get ('/b', (res, req) (res, req)=>{
         runMiddlewares(res,req);  first and second middle to run
     }). get ('/c', (res, req) (res, req)=>{
         runMiddlewares(res,req); first, second and third middle to run
     })

How to use **_enhanceRes_** to implement:  the first middleware to run in  ‘/a’  router ; first and second middle to run in /b
router; and  first, second and third middle to run in /c router

thank you
